[16.0][OU-ADD] account: Migration

ngochung207 opened the pull request:
Video test localhost:
https://youtu.be/f1vUk11DpiU

ngochung207 committed:
[OU-ADD] accont: Migration to 16

account: more analysis on xml
account.accouunt: Account type in account template and account
 - account.move: auto_post
 - account.move.line: Delete sql_contraints, display_type
 - account.move.line: Dynamic analytic_distribution fill yeahhh
 - account.journal: Delete sql_contraints
 - account.payment: Add field amount_company_currency_signed
 - account.bank.statement: internal_index, first_line_index, is_complete

daiduongnguyen-odoo commented on docsource/modules150-160.rst:
```diff
@@ -6,7 +6,7 @@ Module coverage 15.0 -> 16.0
 +-------------------------------------------------+----------------------+-------------------------------------------------+
 | Module                                          | Status               + Extra Information                               |
 +=================================================+======================+=================================================+
-| account                                         |                      |                                                 |
```
Partial everybody :v

daiduongnguyen-odoo replied:
Todo: a dynamic way to fill `anaylytic_distribution` on `account.move.line`, `sale.order.line` for short is any model that inherit from `analytic.mixin`

legalsylvain commented:
/ocabot migration account


<dependency>Depends on : 

- analytic : #4052</dependency>

legalsylvain commented:
@remytms could you take a look here ?

daiduongnguyen-odoo commented:
> @remytms could you take a look here ?

This PR haven't ready yet , but still me and my friend welcome all reviews from anybody especially on the `analysis file`

daiduongnguyen-odoo commented:
Dependency:

- [ ] https://github.com/OCA/OpenUpgrade/pull/4052

legalsylvain commented:
@duong77476. No need to write dependencies message. It will be done automatically by this script. (and the list will be updated regularly). https://github.com/OCA/OpenUpgrade/issues/3979
thanks ! 
Note : You can put your PR in a draft state when the work is not ready to review.

daiduongnguyen-odoo commented:
> @duong77476. No need to write dependencies message. It will be done automatically by this script. (and the list will be updated regularly). #3979 thanks ! Note : You can put your PR in a draft state when the work is not ready to review.

Okay, thanks, just a normal habit of me when a PR need to depend on something, cool tool btw keep it up

legalsylvain commented:
> cool tool btw keep it up

thanks ! 
Well, when there are 100 PR per week (!), It's starting to become mandatory to be able to find your way around.

daiduongnguyen-odoo commented:
@legalsylvain Do you have tools that allow to rebase automatically, as many PRs are coming , i always have to do it manually :((

legalsylvain commented:
> @legalsylvain Do you have tools that allow to rebase automatically, as many PRs are coming , i always have to do it manually :((

You can call ``ocabot rebase``  (with ``\`` before.)

If you have write access on the branch of the PR, it should work. Can you try ?

daiduongnguyen-odoo commented:
\ocabot 

> ocabot rebase

Not working.
IMO, the person who has write access to his/her PR should be able to call that

ngochung207 commented:
@OCA-git-bot rebase

ngochung207 commented:
/ocabot rebase

pedrobaeza commented:
Again you are rebasing over an outdated remote and bringing here lots of commits. Please fetch the remotes to let only the applicable commits.

daiduongnguyen-odoo commented:
> Again you are rebasing over an outdated remote and bringing here lots of commits. Please fetch the remotes to let only the applicable commits.

Sorry about that :((

pedrobaeza commented:
Thank you very much for fixing it.

jcdrubay commented on openupgrade_scripts/scripts/account/16.0.1.2/pre-migration.py:
```diff
@@ -0,0 +1,632 @@
+@openupgrade.migrate()
+def migrate(env, version):
+    # Perform rename stuff
+    openupgrade.rename_xmlids(env, _xmlids_renames)
```
```suggestion
    openupgrade.rename_xmlids(env.cr, _xmlids_renames)
```

Otherwise, I am getting this error:

```
site-packages/openupgradelib/openupgrade.py", line 1236, in rename_xmlids
    cr.execute(get_data_query, tuple(old.split(".")))
    ^^^^^^^^^^
AttributeError: 'Environment' object has no attribute 'execute'
```

https://github.com/OCA/openupgradelib/blob/master/openupgradelib/openupgrade.py#L1211

The code of the latest version of openupgradelib is like this:

![image](https://github.com/OCA/OpenUpgrade/assets/1765420/879bc25c-b66f-4299-bb4c-0224024ae998)

daiduongnguyen-odoo replied:
Sorry about that, i have fixed

pedrobaeza commented:
Again you have pushed a lot of commits due to an outdated branch over which you have rebased.

daiduongnguyen-odoo commented:
> Again you have pushed a lot of commits due to an outdated branch over which you have rebased.

Calm down sir, it is under my control :v

pedrobaeza commented:
Take into account that each time you push against the branch of the pull request, you trigger 2 GitHub actions builds, wasting resources for nothing. Let's be the most ecologic-friendly!

daiduongnguyen-odoo commented:
> Take into account that each time you push against the branch of the pull request, you trigger 2 GitHub actions builds, wasting resources for nothing. Let's be the most ecologic-friendly!

Sorry sir, won't happen again :((

jcdrubay commented:
For the record, I could run the upgrade of a database which is focussed on accounting only.  👍 

I only had to do a tiny adjustments possibly because you are not using the latest version of OpenUpgradeLib when testing.

I will test more in coming days, it's very promising. Thanks for the hard work  @ngochung207 @duong77476 

Thanks @pedrobaeza for ensure that the process is always respected, as well as the environment 😄

daiduongnguyen-odoo commented:
> For the record, I could run the upgrade of a database which is focussed on accounting only. +1
> 
> I only had to do a tiny adjustments possibly because you are not using the latest version of OpenUpgradeLib when testing.
> 
> I will test more in coming days, it's very promising. Thanks for the hard work @ngochung207 @duong77476
> 
> Thanks @pedrobaeza for ensure that the process is always respected, as well as the environment smile

thanks a lot for the test sir, let's us know if some bug or some strange thing occur

daiduongnguyen-odoo committed:
[OU-IMP] account: simplify some fast fill stuff and remove unneeded
method

-display_type improvement
-remove bank statement fastfill
-add account_reconcile_model_line analytic_distribution generate
-improve analytic_distribution_model generation from analytic default

chrisandrewmann commented:
Hi @ngochung207 @duong77476
I'm very new to OpenUpgrade but have just attempted an upgrade using a DB that has account + analytic installed (using this and latest PRs for each script combined).

Error thrown which seems to be related to a view that is no longer present in V16 (just checked and it was in V15).

odoo.tools.convert.ParseError: while parsing file:/c:/odoodev/server-v16/addons/account/report/account_invoice_report_view.xml:90
Error while validating view near:

<search string="Invoices Analysis" __validate__="1">
                <filter string="My Invoices" name="my_invoice" domain="[('invoice_user_id', '=', uid)]"/>
                <separator/>
                <field name="invoice_date"/>
                <separator/>

Unknown field "analytic_account_id" in "group_by" value in context="{'group_by':'analytic_account_id'}"

View error context:
{'file': 'c:\\odoodev\\server-v16\\addons\\account\\report\\account_invoice_report_view.xml',
 'line': 3,
 'name': 'account.invoice.report.search',
 'view': ir.ui.view(1999,),
 'view.model': 'account.invoice.report',
 'view.parent': ir.ui.view(),
 'xmlid': 'view_account_invoice_report_search'}

2023-08-21 13:21:10,566 13412 CRITICAL odoo.service.server: Failed to initialize database. 
Traceback (most recent call last):
  File "c:\odoodev\server-v16\odoo\service\server.py", line 1299, in preload_registries
    registry = Registry.new(dbname, update_module=update_module)
  File "<decorator-gen-14>", line 2, in new
  File "c:\odoodev\server-v16\odoo\tools\func.py", line 87, in locked
    return func(inst, *args, **kwargs)
  File "c:\odoodev\server-v16\odoo\modules\registry.py", line 90, in new
    odoo.modules.load_modules(registry, force_demo, status, update_module)
  File "c:\odoodev\server-v16\odoo\modules\loading.py", line 484, in load_modules
    processed_modules += load_marked_modules(cr, graph,
  File "c:\odoodev\server-v16\odoo\modules\loading.py", line 372, in load_marked_modules
    loaded, processed = load_module_graph(
  File "c:\odoodev\server-v16\odoo\modules\loading.py", line 231, in load_module_graph
    load_data(cr, idref, mode, kind='data', package=package)
  File "c:\odoodev\server-v16\odoo\modules\loading.py", line 71, in load_data
    tools.convert_file(cr, package.name, filename, idref, mode, noupdate, kind)
  File "c:\odoodev\server-v16\odoo\tools\convert.py", line 763, in convert_file
    convert_xml_import(cr, module, fp, idref, mode, noupdate)
  File "c:\odoodev\server-v16\odoo\tools\convert.py", line 829, in convert_xml_import
    obj.parse(doc.getroot())
  File "c:\odoodev\server-v16\odoo\tools\convert.py", line 749, in parse
    self._tag_root(de)
  File "c:\odoodev\server-v16\odoo\tools\convert.py", line 698, in _tag_root
    f(rec)
  File "c:\odoodev\server-v16\odoo\tools\convert.py", line 709, in _tag_root
    raise ParseError(msg) from None  # Restart with "--log-handler odoo.tools.convert:DEBUG" for complete traceback
odoo.tools.convert.ParseError: while parsing file:/c:/odoodev/server-v16/addons/account/report/account_invoice_report_view.xml:90
Error while validating view near:

<search string="Invoices Analysis" __validate__="1">
                <filter string="My Invoices" name="my_invoice" domain="[('invoice_user_id', '=', uid)]"/>
                <separator/>
                <field name="invoice_date"/>
                <separator/>

Unknown field "analytic_account_id" in "group_by" value in context="{'group_by':'analytic_account_id'}"

View error context:
{'file': 'c:\\odoodev\\server-v16\\addons\\account\\report\\account_invoice_report_view.xml',
 'line': 3,
 'name': 'account.invoice.report.search',
 'view': ir.ui.view(1999,),
 'view.model': 'account.invoice.report',
 'view.parent': ir.ui.view(),
 'xmlid': 'view_account_invoice_report_search'}

daiduongnguyen-odoo commented:
Thanks @chrisandrewmann . But as you can see someone else has attempted and worked. If you could provide a video , show me your odoo conf as well, i might be  able to check it out. Otherwise , i really hard for me though

chrisandrewmann commented:
Apologies @duong77476 my mistake.
Was not loading openupgrade framework correctly in conf! Working well now.

daiduongnguyen-odoo commented:
> Apologies @duong77476 my mistake. Was not loading openupgrade framework correctly in conf! Working well now.

Thank you very much, let's me know if something wrong or need improvement.

sbidoul commented on openupgrade_scripts/scripts/account/16.0.1.2/pre-migration.py:
```diff
@@ -0,0 +1,563 @@
+            SET journal_id = aml.journal_id
+        FROM account_move_line aml
+        WHERE aml.id = aal.move_line_id
+        """,
```
I got `insert or update on table "account_analytic_line" violates foreign key constraint "account_analytic_line_journal_id_fkey"` on this update. 

We should probably drop the foreign key constraint before applying this:
`ALTER TABLE account_analytic_line DROP CONSTRAINT IF EXISTS account_analytic_line_journal_id_fkey;`

Because that constraints from v15 and earlier relates to analytic journals.

sbidoul replied:
Another issue I've encountered around here is with analytic lines linked to the Timesheets Journal, which prevent the journal_id foreign key to apply. I've set journal_id to NULL on these records as a workaround. Not sure where the best place to do that is.

remi-filament commented on openupgrade_scripts/scripts/account/16.0.1.2/pre-migration.py:
```diff
@@ -0,0 +1,563 @@
+from openupgradelib import openupgrade
+
+_xmlids_renames = [
```
Hi @ngochung207 Thanks for your work !
I just tested it and found out that this rename is not done (since not called in migrate function below AFAICT).

You may want to add a `openupgrade.rename_xmlids(env.cr, _xmlids_renames)` to enforce it !

daiduongnguyen-odoo replied:
I know, i have postpone the work from this PR, you may want to try at https://github.com/Viindoo/OpenUpgrade , my work has finished there

remi-filament replied:
Thanks @duong77476 I manage to fully migrate my database using both your repo and PRs from OCA/OpenUpgrade repo. It is however very difficult to track differences since this PR and your repo contain a lot of things which are related to your instances / modules.
Could you please let me know your plans to complete/cleanup this PR ? I can help if needed !
Best Regards

ahshakir commented:
Hi,
I have been using OpenUpgrade to migrate from version 11 to version 16. So far till version 15 everything works fine and I know the migration script for 16 is not completely ready yet but I have to complete this migration so I collected the migration scripts from different repos to use and for account module I used script from here. Now the problem is after migration I tried to view the invoices and the invoice lines are not being displayed for some reason. Kindly advise please to rectify this. Thank you in advance.

daiduongnguyen-odoo commented:
> Hi, I have been using OpenUpgrade to migrate from version 11 to version 16. So far till version 15 everything works fine and I know the migration script for 16 is not completely ready yet but I have to complete this migration so I collected the migration scripts from different repos to use and for account module I used script from here. Now the problem is after migration I tried to view the invoices and the invoice lines are not being displayed for some reason. Kindly advise please to rectify this. Thank you in advance.

Thanks, you may want to try at https://github.com/Viindoo/OpenUpgrade instead

ahshakir commented:
> > Hi, I have been using OpenUpgrade to migrate from version 11 to version 16. So far till version 15 everything works fine and I know the migration script for 16 is not completely ready yet but I have to complete this migration so I collected the migration scripts from different repos to use and for account module I used script from here. Now the problem is after migration I tried to view the invoices and the invoice lines are not being displayed for some reason. Kindly advise please to rectify this. Thank you in advance.
> 
> Thanks, you may want to try at https://github.com/Viindoo/OpenUpgrade instead

I tried them too but still it didn't work as expected. Also did you come across the same error as mine after migration? And do you think is there any other way possibly I could try to fix this issue please. I am in real need of this fix right now.

daiduongnguyen-odoo commented:
@ahshakir no unfortunately, do you have any custom module beside odoo module ?

ahshakir commented:
> @ahshakir no unfortunately, do you have any custom module beside odoo module ?

Yes I do and I upgraded them too. Though some modules created new fields in account but nothing sort of doing major change in them And all other modules working fine except for account.

daiduongnguyen-odoo commented:
@ahshakir OK now go to debug mode, check the `Edit form view` then see the view architecture of account_move to see anything strange

ahshakir commented:
I did not see anything strange. but what I would like to add is 
The table names in account in Odoo 11, 
account invoice - Invoices
account invoice line - Invoice Lines 
account move - Journal Entry 
account move line - Journal Items 
and from Odoo 13, Invoices had name changed from account invoice to account move and from there Invoices, Journal Entries everything feeds into same table if I am not wrong.
Now coming to version 16 Invoices and Journal Entries are in account move and in the db also everything is in the place but still unable to view them. Also there is field in move line I saw display_type. does it have anything to do with the invoices?

ikus060 commented:
@duong77476 I wish to test your migration script, but I'm block by the migration of "payment". Do you happen to have a branch somewhere with both of these changes ?

As recommended, I already tried to upgrade by using https://github.com/Viindoo/OpenUpgrade But it it's failing the same way.

daiduongnguyen-odoo commented:
> @duong77476 I wish to test your migration script, but I'm block by the migration of "payment". Do you happen to have a branch somewhere with both of these changes ?
> 
> As recommended, I already tried to upgrade by using https://github.com/Viindoo/OpenUpgrade But it it's failing the same way.

What is blocking exactly , show me error on your console log if you have any ? i have tried on a production database and it works fine

ikus060 commented:
Here the relevant part where the migration is failing. But it's not related to "account" migration. It's the "payment" migration that is failing.

[odoo.log](https://github.com/OCA/OpenUpgrade/files/12894715/odoo.log)

Thanks for your help.

duong77476 commented:
> Here the relevant part where the migration is failing. But it's not related to "account" migration. It's the "payment" migration that is failing.
> 
> [odoo.log](https://github.com/OCA/OpenUpgrade/files/12894715/odoo.log)
> 
> Thanks for your help.

Are you sure i have correct odoo 16.0 addons when running migration, i notice that in odoo 16 CE , in folder `data`  it is kind of different compare with what i have seen in your log. 
https://github.com/odoo/odoo/blob/16.0/addons/payment_aps/data/payment_provider_data.xml

daiduongnguyen-odoo commented:
> I did not see anything strange. but what I would like to add is The table names in account in Odoo 11, account invoice - Invoices account invoice line - Invoice Lines account move - Journal Entry account move line - Journal Items and from Odoo 13, Invoices had name changed from account invoice to account move and from there Invoices, Journal Entries everything feeds into same table if I am not wrong. Now coming to version 16 Invoices and Journal Entries are in account move and in the db also everything is in the place but still unable to view them. Also there is field in move line I saw display_type. does it have anything to do with the invoices?

Sorry, it is very complex when running migration all the way up from 11 to 16. I have ran it from 10 to 14 already, i must tell you that is not an easy thing to do. You will have to make sure each migration jump (11 -> 12, 12->13, ...) work smootly

daiduongnguyen-odoo commented:
> Here the relevant part where the migration is failing. But it's not related to "account" migration. It's the "payment" migration that is failing.
> 
> [odoo.log](https://github.com/OCA/OpenUpgrade/files/12894715/odoo.log)
> 
> Thanks for your help.

Please , check your odoo 16 CE addons, why the log said `base.module_payment_abs` not found, strange

ikus060 commented:
> Please , check your odoo 16 CE addons, why the log said `base.module_payment_aps` not found, strange

The module `payment_aps` seams to be a new module in Odoo v16.0. My database doesn't have any entry in table `ir_module_module` for this new module. Not sure if a script should be responsible to create an entry for it !?

ahshakir commented:
> > I did not see anything strange. but what I would like to add is The table names in account in Odoo 11, account invoice - Invoices account invoice line - Invoice Lines account move - Journal Entry account move line - Journal Items and from Odoo 13, Invoices had name changed from account invoice to account move and from there Invoices, Journal Entries everything feeds into same table if I am not wrong. Now coming to version 16 Invoices and Journal Entries are in account move and in the db also everything is in the place but still unable to view them. Also there is field in move line I saw display_type. does it have anything to do with the invoices?
> 
> Sorry, it is very complex when running migration all the way up from 11 to 16. I have ran it from 10 to 14 already, i must tell you that is not an easy thing to do. You will have to make sure each migration jump (11 -> 12, 12->13, ...) work smootly

I didn't check thoroughly with everything working smoothly while migrating to every version but I do checked if the invoices were visible properly in every migration and they were there. Only in version 16 they weren't visible at all. ALSO compared the columns in version 15 and 16 of table account.move and didn't find anything different between them. 

Is there anyone here migrated to 16 and came across something similar to what I found??

daiduongnguyen-odoo commented:
> > Please , check your odoo 16 CE addons, why the log said `base.module_payment_aps` not found, strange
> 
> The module `payment_aps` seams to be a new module in Odoo v16.0. My database doesn't have any entry in table `ir_module_module` for this new module. Not sure if a script should be responsible to create an entry for it !?

No, i dont think because of the script, here is the config that i have used to run migration script. I use eclipse as an IDE to run this, of course you will need to make sure both odoo 15 and odoo 16 CE has the latest code
--addons-path=~/git/odoo16/addons,~/git/Openupgrade16
--load=base,web,openupgrade_framework
--limit-time-real=10000000
--http-port=8016
-d oca1
-u all
--upgrade-path=~/git/Openupgrade16/openupgrade_scripts/scripts

daiduongnguyen-odoo commented:
> > > I did not see anything strange. but what I would like to add is The table names in account in Odoo 11, account invoice - Invoices account invoice line - Invoice Lines account move - Journal Entry account move line - Journal Items and from Odoo 13, Invoices had name changed from account invoice to account move and from there Invoices, Journal Entries everything feeds into same table if I am not wrong. Now coming to version 16 Invoices and Journal Entries are in account move and in the db also everything is in the place but still unable to view them. Also there is field in move line I saw display_type. does it have anything to do with the invoices?
> > 
> > 
> > Sorry, it is very complex when running migration all the way up from 11 to 16. I have ran it from 10 to 14 already, i must tell you that is not an easy thing to do. You will have to make sure each migration jump (11 -> 12, 12->13, ...) work smootly
> 
> I didn't check thoroughly with everything working smoothly while migrating to every version but I do checked if the invoices were visible properly in every migration and they were there. Only in version 16 they weren't visible at all. ALSO compared the columns in version 15 and 16 of table account.move and didn't find anything different between them.
> 
> Is there anyone here migrated to 16 and came across something similar to what I found??

Maybe try `inspect (f12)` to see if it still there. You mean `not visible` here is the tab `Invoice Line` is missing right ?

pedrobaeza committed:
[OU-FIX+IMP] account: Refine migration scripts

- Don't load non useful noupdate_changes
- Precompute account.payment~amount_company_currency_signed for the
  easy case of same currency.
- Preserve old account.move~auto_post column.
- Adapt account.account~account_type filling SQL and add a fallback.
- Fill properly account.account~include_initial_balance.
- No need of fast filling account.analytic.line~journal_id.
- Adapt account_analytic_distribution_model generation SQL.
- Adapt analytic_distribution SQL filling and make it generic.

 TT42204

pedrobaeza commented:
Thanks to the Viindoo pioneers on this difficult task.

I have performed the following changes and push in this PR (as there are permissions to maintainers granted):

- Don't load non useful noupdate_changes.
- Precompute account.payment~amount_company_currency_signed for the easy case of same currency.
- Preserve old account.move~auto_post column.
- Adapt account.account~account_type filling SQL and add a fallback.
- Fill properly account.account~include_initial_balance.
- No need of fast filling account.analytic.line~journal_id.
- Adapt account_analytic_distribution_model generation SQL.
- Adapt analytic_distribution SQL filling and make it generic.

I'm merging as soon as this is green, so finally! :tada:

daiduongnguyen-odoo commented:
> Thanks to the Viindoo pioneers on this difficult task.
> 
> I have performed the following changes and push in this PR (as there are permissions to maintainers granted):
> 
> * Don't load non useful noupdate_changes.
> * Precompute account.payment~amount_company_currency_signed for the easy case of same currency.
> * Preserve old account.move~auto_post column.
> * Adapt account.account~account_type filling SQL and add a fallback.
> * Fill properly account.account~include_initial_balance.
> * No need of fast filling account.analytic.line~journal_id.
> * Adapt account_analytic_distribution_model generation SQL.
> * Adapt analytic_distribution SQL filling and make it generic.
> 
> I'm merging as soon as this is green, so finally! 🎉

At your service sir :hear_no_evil:
Thank you for your work too
